### PR TITLE
Profile page

### PR DESCRIPTION
**Pull request by bxchang04**
I was able to implement a profile page with 2 fields -- one for firstname, one for lastname.

At first I was also able to get the "update profile" button to create a new db entry in the table. However due to some GIT issues and my personal error, I lost those changes, and this version of the code does not successfully create a new table row on click.

However, despite not being able to recover that functionality, it's not according to spec anyways. For the next step, I would like to have the user's default firstname and lastname appear automatically on the 'Profile' page. However in order for this to be possible, I assume the register feature has to be implemented first.

Next steps: Would it be worth it fixing the broken 'Update Profile' button as-is, or for me to wait on the register feature to be completed before attempting to load up a default 'firstname' and 'lastname', and completing the modify firstname / lastname feature for this issue?

Also, please let me know if I'm on the right track, and if my thought process above makes sense. I also thought about mocking up a default firstname/lastname by using the login.js file to set a temporary firstname/lastname, but got overwhelmed by what that requires. Looking forward to your feedback, and thanks for letting me attempt this issue!

### CODE REVIEW

**Comment by alexjball** [comment]
What do you mean by default first/last name? The first/last name set when the account is created? I was thinking that if the user doesn't have first/last name set, we render empty form fields. Then, to update first/last name, we require both. This will allow us to use the profile page as the main editor for profile information and keep the initial register feature minimal

**Comment by bxchang04** [comment]
Yes you guessed correctly -- in HTML, default is what renders when the field is blank.

We could implement as you suggested above, which I assume is to leave the default for both fields as blank. It doesn't seem as elegant, but it's definitely less complicated since we don't have to get anything from Firebase for a brand new user. I was able to get a firstname and lastname to be created on click of 'update profile', but i lost that code and now nothing is created when I click update.

So in other words from what I understand, what we want is for the same entry to be created or updated each time the user clicks "Update Profile." Currently the user's firebase entry only has login email and password....  I assume we would need to add "firstname" and "lastname" as columns as well? Could you give me any tips on how to refactor my current code to accomplish this? I'm afraid I don't know Firebase well enough to code this independently.

File: src/ProfileTable.js [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+const column = (name, getter) => ({ name, getter });
+const columns = [
+  column('First Name', row => row.firstName),
+  column('Last Name', row => row.lastName),
+];
+
+const fullName = users =>
+  `${users.firstName} ${users.lastName}`.trim();
+
+export default () => {
```
Comment (alexjball) [inline src/ProfileTable.js]: As you said in your comment, we will only be reading and updating the user document for the current users, so we don't need this table

File: src/RecordProfileForm.js~Stashed changes [hunk 2]
```diff
@@ -0,0 +1,79 @@
+import React, { useReducer } from 'react';
```
Comment (alexjball) [inline src/RecordProfileForm.js~Stashed changes]: Please remove this file, it looks like it was generated by git

File: src/TopBar.js [hunk 3]
```diff
@@ -15,11 +15,14 @@ export default () => {
         <Item to="/incidents" label="Incidents" />
         <Item to="/engagements" label="Engagements" />
         <div style={{ flexGrow: 1 }} />
+        <Button variant="dark" label="Profile" style={{ color: "black" }}  > {/* changing font color to dark doesn't work here for some reason, so I changed the button from light themed to dark themed for now */}
```
Comment (alexjball) [inline src/TopBar.js]: Currently, the user needs to click the text to navigate to the profile page. Since this is a button, it should work to click anywhere inside the outline. Check out this example for how to use the onClick callback to navigate with history.push('/profile'): https://reacttraining.com/react-router/web/api/Hooks/usehistory

File: src/RecordProfileForm.js [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+export default () => {
+  const [state, update] = useReducer(
+    (state, update) => ({ ...state, ...update }),
+    {
+      firstName: '',
+      lastName: '',
+    }
+  );
+
```
Comment (alexjball) [inline src/RecordProfileForm.js:15]: You can load the current document like this:
```
const { user } = useAuthState()
const userDoc = db.collection('users').doc(user.uid)

useEffect(() => {
  const loadUser = async () => {
    const {firstName, lastName} = await userDoc.read();
    update({firstName, lastName});
  }
  loadUser();
}, [])
```

File: src/RecordProfileForm.js [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    const id = e.target.id;
+    update({ [id]: value });
+  };
+
+  const handleSubmit = async event => {
+    event.preventDefault();
+    event.stopPropagation();
+
+    try {
+      await db.collection('participants').add({
```
Comment (alexjball) [inline src/RecordProfileForm.js]: Here you want to update the user you read from earlier rather than add to the table:

`await userDoc.update({firstName, lastName})`

you can test that this works by refreshing the page. The name should reflect the updated value.

**Review by alexjball — COMMENTED** [review]
Great progress Ben! Happy to walk through this on Tuesday!

**Comment by alexjball** [comment]
Hey Ben! Thank you for contributing! Just clarifying some things below and listing expected behaviors

There are two "users" here: one is the Firebase auth user we get on login and access with useAuthState, and the other are documents in the "users" table. The auth user is stored for us by Google, so it is minimal and can only contain the email, password, and a unique ID. We own the users table and use it to store any information that is specific to an individual user, such as first and last name. We store users in the users table by the auth user's ID, which allows us to look up information for the currently logged in user.

- Form fields should initialize to the values in the current user doc, or empty if there aren't currently first/last names set.
- Clicking the update profile button should update the current user doc with the values in the form fields.
- The form fields are required, meaning that users can't update to empty fields.

Since we don't have a registration flow yet, users will use the profile page to initially input their name, so it's important we support the empty fields. As you say, in the future, the registration form can initialize the user document with first/last name, and the profile page will immediately reflect the user's name on login.

File: src/RecordProfileForm.js (hunk 4 above)
Comment (bxchang04) [reply src/RecordProfileForm.js:15]: Got it. Does this mean I need to overwrite lines 12-15, or to inject this code into the parenthesis, squiggly bracket? 

Not quite sure how this all works together as I'm still learning the structure and syntax here..

File: src/TopBar.js (hunk 3 above)
Comment (bxchang04) [reply src/TopBar.js]: Added a handleclick function to this button... but haven't tested this out yet. May I get a password reset? (Doh!)

File: src/ProfileTable.js (hunk 1 above)
Comment (bxchang04) [reply src/ProfileTable.js]: Will delete this entire file eventually. Still need to test the code in the files below.

File: src/RecordProfileForm.js (hunk 5 above)
Comment (bxchang04) [reply src/RecordProfileForm.js]: Haven't tested this yet due to issues with comment above.

File: src/TopBar.js (hunk 3 above)
Comment (bxchang04) [reply src/TopBar.js]: works!

File: src/RecordProfileForm.js (hunk 5 above)
Comment (bxchang04) [reply src/RecordProfileForm.js]: works now :)

File: src/RecordProfileForm.js (hunk 4 above)
Comment (bxchang04) [reply src/RecordProfileForm.js:15]: fixed during tonight's pair coding session

**Comment by bxchang04** [comment]
Now that everything's addressed, I will submit a new pull request with the latest code.

File: src/RecordProfileForm.js (hunk 4 above)
Comment (bxchang04) [reply src/RecordProfileForm.js:15]: fixed during tonight's pair coding session
